Add mod to redirect Modern Notepad to Classic Editor

CherryDT opened the pull request:
This mod redirects Windows 11's Notepad to a configurable classic editor, including an elevate-on-Ctrl+Shift feature.

m417z commented:
I tried the mod, and it doesn't work for me. It causes the modern Notepad to be launched again and again, resulting in a huge amount of Notepad windows after a single launch.

I tried to implement a similar mod at some point, but came to the conclusion that the redirection functionality is deeply integrated into Windows, and it makes more sense to just disable the modern Notepad via the settings, under "App execution aliases".

CherryDT commented:
This mod goes beyond what disabling the alias can do, but I forgot that the app execution alias needs to be disabled too for this to work. I now added a check for it and if the user agrees I automatically disable the alias.

CherryDT commented:
Btw, in case you wonder why I did force-push all the time - sorry, it was a misunderstanding, I just realized. When I read the readme, somehow I read the PR has to consist of a single _commit_, when in fact it just said to contain a single _file_.

CherryDT commented:
I just realized (through testing) that `%programfiles%` _does_ work in `@include`, I didn't find this in the docs. Would be nice to add it because I think in general it should be preferred not to hardcode such paths!

m417z commented on mods/modern-notepad-to-classic-editor.wh.cpp:
```diff
@@ -0,0 +1,429 @@
+BOOL Wh_ModInit() {
+  // Load settings
+  {
+    std::wstring pathSetting = Wh_GetStringSetting(L"classicPath");
```
There's a memory leak here. You can use `WindhawkUtils::StringSetting` for a C++ wrapper that frees it automatically.

CherryDT replied:
Fixed

m417z commented on mods/modern-notepad-to-classic-editor.wh.cpp:
```diff
@@ -0,0 +1,429 @@
+
+  // Only act inside the WindowsApps Notepad process
+  std::wstring cur = GetCurrentExePath();
+  if (!IsModernWindowsAppsNotepadPath(cur)) {
```
Isn't this check redundant since the mod only targets the modern Notepad already?

CherryDT replied:
Yes, at some point I had just `notepad.exe` because I didn't want to hardcode the program files path and didn't realize environment variables work there.

m417z commented on mods/modern-notepad-to-classic-editor.wh.cpp:
```diff
@@ -0,0 +1,429 @@
+  }
+
+  // If this is already an "inner" launch with /SESSION, exit quietly
+  if (ArgsContainSession(argv)) {
```
You can use `isSessionLaunch` here.

m417z commented on mods/modern-notepad-to-classic-editor.wh.cpp:
```diff
@@ -0,0 +1,429 @@
+        if (val.find(L"WindowsApps") != std::wstring::npos) {
+          int result = isSessionLaunch ? IDNO : MessageBoxW(NULL, L"It seems like an app execution alias for notepad.exe to the modern Notepad is set in Windows settings and this mod cannot work while the alias is enabled. Do you want to disable this alias?", L"Redirect Modern Notepad to Classic Editor (Windhawk Mod)", MB_YESNO | MB_ICONQUESTION);
+          if (result == IDYES) {
+            RegCloseKey(hKey);
```
Here, `RegCloseKey(hKey)` will be called twice.

CherryDT replied:
Fixed.

m417z commented on mods/modern-notepad-to-classic-editor.wh.cpp:
```diff
@@ -0,0 +1,465 @@
+
+    if (ShellExecuteExW(&sei)) {
+      if (g_waitForClassic && sei.hProcess) {
+        WaitForSingleObject(sei.hProcess, INFINITE);
```
This case, and the `WaitForSingleObject` below, cause the hourglass/loading cursor to be shown for some time. It can be avoided with [something like this](https://stackoverflow.com/a/3865519).

CherryDT commented:
Thanks, implemented all your suggestions.

Also removed the `ms-notepad:` URI handling because regular immersive notepad doesn't handle it either (and I was not handling encoded spaces anyway).
